model/Command/ChangePermissionsCommand.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2023 (original work) Open Assessment Technologies SA.
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoDacSimple\model\Command;
+
+use core_kernel_classes_Resource;
+
+/**
+ * Value object holding data about permission changes to be done in a resource.
+ *
+ * By default, the command is set up to apply a given set of ACLs to a single
+ * resource designated as the command's root.
+ *
+ * However, it also provides methods to create copies of the command having
+ * different recursion settings, making PermissionsService to operate
+ * recursively or to modify class instances under the root resource in case the
+ * root itself is a class.
+ */
+class ChangePermissionsCommand
+{
+    private core_kernel_classes_Resource $root;
+
+    /**
+     * An array in the form ['userId' => ['READ', 'WRITE], ...]
+     *
+     * @var string[]
+     */
+    private array $privilegesPerUser;
+
+    private bool $isRecursive = false;
+
+    private bool $applyToNestedResources = false;
+
+    public function __construct(
+        core_kernel_classes_Resource $root,
+        array $privileges
+    ) {
+        $this->root = $root;
+        $this->privilegesPerUser = $privileges;
+    }
+
+    /**
+     * Clones the current command setting its recursive flag to a new value.
+     *
+     * For commands having a class as the root, setting the recursion flag
+     * makes PermissionsService to update permissions for the class and all its
+     * descendants (i.e. updates all resources AND classes using the provided
+     * root class as the initial node for a tree traversal, which may be slow
+     * and resource-intensive).
+     *
+     * This is provided for backward compatibility purposes.
+     */
+    public function withRecursion(bool $isRecursive = true): self
```

Comment:
Default is false, this is a "switch" method. Bolean param is considered a code smell https://martinfowler.com/bliki/FlagArgument.html

```suggestion
    public function withRecursion(): self
```

Reply:
I'll put the logic in the caller instead and make the command a mutable object.

_(However, citing Frowler's article, when "pulling data from a boolean source, such as a UI control or data source"--as we are doing here--", I'd rather have `setSwitch(aValue)` than `if (aValue) { setOn(); } else { setOff(); }`)._

Reply:
I prefer be consistent and use the same approach rather than find exceptions. We might use this command not coming from UI, then this premise is not valid. Either way, this is optional. 

To be honest, I do not know why we need those methods instead of passing params by constructor, but it is ok :)

Reply:
1abe4f4f36035340159f4a50e8627d48ae5e9af4

Reply:
I thought it will be easier to understand something like 

```
(new ChangePermissionsCommand($resource, $privilegesToSet))
    ->withRecursion()
    ->withoutNestedResources()
```
instead of

```
new ChangePermissionsCommand($resource, $privilegesToSet, true, false)
```

because, for the later, someone working on the caller code will need to check the signature of the command constructor to know what each flag parameter means (and then keep remembering in which order the flags were declared in the constructor).

It was also a bit more difficult to follow during the development having the two flag parameters (I had to put comments during the development in front of each one with their meaning in order to not misinterpret them).

I also thought that that won't scale well in case we find a need to keep adding more flag parameters to the command in the future.